### PR TITLE
Standalone Lite: Update LightningLite

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Updates the LightningLite class to make use of the newly added standalone connector, strategies, and other APIs. 
Also contains minor fixes and todo resolutions. 

In the next PR, the class and wrappers will be moved to the standalone package. 
In another PR, we will add a subset of integration tests. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @tchaton @rohitgr7 @justusschock @awaelchli @carmocca @akihironitta

### CODE REVIEW

File: src/pytorch_lightning/lite/lite.py [hunk 1]
```diff
@@ -25,23 +25,23 @@
...
     _auto_add_worker_init_fn,
     _replace_dunder_methods,
     _update_dataloader,
     has_iterable_dataset,
 )
+from lightning_lite.utilities.exceptions import MisconfigurationException
 from lightning_lite.utilities.seed import seed_everything
-from pytorch_lightning.accelerators.accelerator import Accelerator
 from pytorch_lightning.lite.wrappers import _LiteDataLoader, _LiteModule, _LiteOptimizer
 from pytorch_lightning.overrides.distributed import DistributedSamplerWrapper
```
Comment (awaelchli) [inline src/pytorch_lightning/lite/lite.py:44]: TODO: Move this

File: src/pytorch_lightning/lite/lite.py [hunk 2]
```diff
@@ -76,34 +76,23 @@ def __init__(
         devices: Optional[Union[List[int], str, int]] = None,
         num_nodes: int = 1,
         precision: Union[int, str] = 32,
-        plugins: Optional[Union[PLUGIN_INPUT, List[PLUGIN_INPUT]]] = None,
+        plugins: Optional[Union[_PLUGIN_INPUT, List[_PLUGIN_INPUT]]] = None,
         gpus: Optional[Union[List[int], str, int]] = None,
         tpu_cores: Optional[Union[List[int], str, int]] = None,
```
Comment (awaelchli) [inline src/pytorch_lightning/lite/lite.py:81]: Open discussion: 
Consider removing these when moving to standalone lite, while keeping them at the same time for the deprecated class in PL

File: src/lightning_lite/connector.py [hunk 3]
```diff
@@ -69,7 +69,7 @@ class _Connector:
             2. accelerator str
             3. accelerator auto
 
-        B. strategy flag could be:
+        B. strategy flag could be :
```
Comment (awaelchli) [inline src/lightning_lite/connector.py:72]:
```suggestion
        B. strategy flag could be:
```

File: src/lightning_lite/lite.py [hunk 4]
```diff
@@ -1,3 +0,0 @@
-class LightningLite:
```
Comment (awaelchli) [inline src/lightning_lite/lite.py:1]: This file has to be removed, otherwise, when we move the Lite class here in the next PR, it won't show up as a clean file move.

File: src/lightning_lite/strategies/ddp_spawn.py [hunk 5]
```diff
@@ -200,8 +196,7 @@ def _setup_distributed(self) -> None:
     def _get_process_group_backend(self) -> str:
         return self._process_group_backend or get_default_process_group_backend_for_device(self.root_device)
 
-    def _set_world_ranks(self, process_idx: int = 0) -> None:
-        self._local_rank = process_idx
+    def _set_world_ranks(self) -> None:
```
Comment (awaelchli) [inline src/lightning_lite/strategies/ddp_spawn.py:199]: This was a bug introduced in #14675 when applying #11073. The launcher already sets the local_rank, but here we were overwriting it to 0. A failing test was pointing me to it.

File: src/lightning_lite/utilities/distributed.py [hunk 6]
```diff
@@ -262,3 +264,58 @@ def _get_process_group_backend_from_env() -> Optional[str]:
...
+
+    def __len__(self) -> int:
+        return len(self._sampler)
+
+    def reset(self) -> None:
+        """Reset the sampler list in order to get new sampling."""
+        self._sampler_list = list(self._sampler)
+
+
+class DistributedSamplerWrapper(DistributedSampler):
```
Comment (awaelchli) [inline src/lightning_lite/utilities/distributed.py:309]: Forgot to move these two classes in the utilities PR #14492

File: src/lightning_lite/strategies/ddp_spawn.py [hunk 7]
```diff
@@ -114,10 +114,6 @@ def setup_module(self, module: Module) -> Module:
         return DistributedDataParallel(module=module, device_ids=self._determine_ddp_device_ids(), **self._ddp_kwargs)
 
     def module_to_device(self, module: Module) -> None:
-        if self.root_device.type == "cuda":
-            # TODO(lite): This should be handled outside module_to_device, by a call to accelerator.setup_device()
```
Comment (awaelchli) [inline src/lightning_lite/strategies/ddp_spawn.py:118]: Addressed todo: See Lite.to_device

File: src/pytorch_lightning/lite/lite.py [hunk 8]
```diff
@@ -429,67 +412,29 @@ def _move_model_to_device(self, model: nn.Module, optimizers: List[Optimizer]) -
...
-            self._accelerator_connector.is_distributed
+            self._connector.is_distributed
             and not isinstance(dataloader.sampler, DistributedSampler)
             and not has_iterable_dataset(dataloader)
         )
 
     @staticmethod
     def _get_distributed_sampler(dataloader: DataLoader, **kwargs: Any) -> DistributedSampler:
         kwargs.setdefault("seed", int(os.getenv("PL_GLOBAL_SEED", 0)))
+        # TODO(lite): Bring the DistributedSamplerWrapper to Lite package
```
Comment (awaelchli) [inline src/pytorch_lightning/lite/lite.py:423]:
```suggestion
```

File: src/pytorch_lightning/lite/lite.py [hunk 9]
```diff
@@ -429,67 +412,29 @@ def _move_model_to_device(self, model: nn.Module, optimizers: List[Optimizer]) -
...
         return DistributedSamplerWrapper(dataloader.sampler, **kwargs)
 
-    def _check_accelerator_support(self, accelerator: Optional[Union[str, Accelerator]]) -> None:
```
Comment (awaelchli) [inline src/pytorch_lightning/lite/lite.py:442]: These checks are now handled with the Lite-specific (accelerator) connector.

File: src/pytorch_lightning/overrides/distributed.py [hunk 10]
```diff
@@ -109,60 +109,6 @@ def __iter__(self) -> Iterator[List[int]]:
...
-
-    def __len__(self) -> int:
-        return len(self._sampler)
-
-    def reset(self) -> None:
-        """Reset the sampler list in order to get new sampling."""
-        self._sampler_list = list(self._sampler)
-
-
-class DistributedSamplerWrapper(DistributedSampler):
```
Comment (awaelchli) [inline src/pytorch_lightning/overrides/distributed.py:151]: Moved these classes without change (only change was neutralizing the mention of PyTorch Lightning in the docstring)

File: tests/tests_lite/test_lite.py [hunk 11]
```diff
@@ -1,12 +0,0 @@
-from tests_lite.helpers.runif import RunIf
```
Comment (awaelchli) [inline tests/tests_lite/test_lite.py:1]: Will be replaced with actual tests via clean move in next PR

File: tests/tests_pytorch/lite/test_lite.py [hunk 12]
```diff
@@ -48,18 +48,6 @@ def forward(self, x):
...
-def test_unsupported_accelerator():
-    accelerator = "coconut"
-    with pytest.raises(MisconfigurationException, match=f"`accelerator={repr(accelerator)}` is not a valid choice"):
-        EmptyLite(accelerator=accelerator)
-
-
-def test_unsupported_strategy():
-    strategy = "coconut"
-    with pytest.raises(MisconfigurationException, match=f"`strategy={repr(strategy)}` is not a valid choice"):
-        EmptyLite(strategy=strategy)
```
Comment (awaelchli) [inline tests/tests_pytorch/lite/test_lite.py:60]: This is tested already in the tests/tests_lite/test_connector.py

**Review by justusschock — APPROVED** [review]

File: src/lightning_lite/connector.py [hunk 13]
```diff
@@ -184,7 +184,8 @@ def _check_config_and_set_final_flags(
         if strategy is not None and strategy not in self._registered_strategies and not isinstance(strategy, Strategy):
             raise ValueError(
                 f"You selected an invalid strategy name: `strategy={strategy!r}`."
-                f" Available names are: {', '.join(self._registered_strategies)}."
+                " Example choices: ddp, ddp_spawn, deepspeed, dp, ..."
+                " Find a complete list of options in our documentation at https://lightning.ai"
```
Comment (carmocca) [inline src/lightning_lite/connector.py:188]: Don't we have a more specific link than the website root?

File: src/pytorch_lightning/lite/wrappers.py [hunk 14]
```diff
@@ -56,21 +52,20 @@ def optimizer(self) -> Optimizer:
         return self._optimizer
 
     def state_dict(self) -> Dict[str, Tensor]:
-        return self._strategy.optimizer_state(self.optimizer)
+        return self._strategy.get_optimizer_state(self.optimizer)
 
-    def step(self, closure: Optional[Callable] = None) -> Any:
-        closure = closure or _do_nothing_closure
+    def step(self, closure: Optional[Callable] = None, module: Optional["_LiteModule"] = None) -> Any:
```
Comment (carmocca) [inline src/pytorch_lightning/lite/wrappers.py:57]: Why do you add this argument? For deepspeed users using Lite?

**Review by carmocca — APPROVED** [review]
:rocket:

File: src/lightning_lite/connector.py (hunk 13 above)
Comment (awaelchli) [reply src/lightning_lite/connector.py:188]: Yes, I could point directly to the Lite page. Are we ok with the risk of it getting outdated?

File: src/pytorch_lightning/lite/wrappers.py (hunk 14 above)
Comment (awaelchli) [reply src/pytorch_lightning/lite/wrappers.py:57]: Yes. It could stay or go, either way it does not make a big difference.

File: src/lightning_lite/utilities/distributed.py [hunk 15]
```diff
@@ -262,3 +263,58 @@ def _get_process_group_backend_from_env() -> Optional[str]:
             " Specify `process_group_backend` directly on the strategy constructor."
         )
     return torch_backend
+
+
+# TODO(lite): The error messsages refer to 'replace_sampler_ddp' in PL but Lite has it named 'replace_sampler'
```
Comment (awaelchli) [inline src/lightning_lite/utilities/distributed.py:268]:
```suggestion
# TODO(lite): The error messages refer to 'replace_sampler_ddp' in PL but Lite has it named 'replace_sampler'
```

File: src/lightning_lite/connector.py (hunk 13 above)
Comment (carmocca) [reply src/lightning_lite/connector.py:188]: You are right, with all the docs changes going on, it's probably better not to do it yet

File: src/pytorch_lightning/lite/wrappers.py (hunk 14 above)
Comment (carmocca) [reply src/pytorch_lightning/lite/wrappers.py:57]: I'd remove it unless there's a real worthy usecase for it

File: src/lightning_lite/connector.py (hunk 13 above)
Comment (tchaton) [reply src/lightning_lite/connector.py:188]: Yes, let's take the risk.

File: src/pytorch_lightning/lite/lite.py [hunk 16]
```diff
@@ -76,34 +75,23 @@ def __init__(
         devices: Optional[Union[List[int], str, int]] = None,
         num_nodes: int = 1,
         precision: Union[int, str] = 32,
-        plugins: Optional[Union[PLUGIN_INPUT, List[PLUGIN_INPUT]]] = None,
+        plugins: Optional[Union[_PLUGIN_INPUT, List[_PLUGIN_INPUT]]] = None,
         gpus: Optional[Union[List[int], str, int]] = None,
         tpu_cores: Optional[Union[List[int], str, int]] = None,
```
Comment (tchaton) [inline src/pytorch_lightning/lite/lite.py:80]: I would advocate to remove them and expose the latest API.

**Review by tchaton — APPROVED** [review]
LGTM !

File: src/pytorch_lightning/lite/lite.py (hunk 16 above)
Comment (awaelchli) [reply src/pytorch_lightning/lite/lite.py:80]: Thanks, sounds good. This likely be done in #14735 or in a follow up.
